app/objects/c_operation.py

```diff
@@ -131,20 +132,27 @@ def add_link(self, link):
     def has_link(self, link_id):
         return any(lnk.id == link_id for lnk in self.potential_links + self.chain)
 
-    def all_facts(self):
-        seeded_facts = [f for f in self.source.facts] if self.source else []
-        learned_facts = [f for lnk in self.chain for f in lnk.facts if f.score > 0]
+    async def all_facts(self):
+        knowledge_svc_handle = BaseService.get_service('knowledge_svc')
+        seeded_facts = []
+        if self.source:
+            seeded_facts = await knowledge_svc_handle.get_facts(criteria=dict(source=self.source.id))
+        learned_facts = await knowledge_svc_handle.get_facts(criteria=dict(source=self.id))
```

Comment:
No, it should also have a `origin_type` of `LEARNED` to denote it was learned. What `self.id` is doing here is that it's helping to narrow down fact results to ones that have originated from this operation (since we track that now). Source seeded facts get the id of the source, facts sent directly to agents (mainly through manual mode, I'm assuming) get the id of the agent, etc.

Reply:
I guess I would have a question on this as well, and looking through the PR it looks like all the places where a paw would be the source it has been replaced by the operation id or the link id. So should source be a collection to allow for both paw and operation id (and whatever else might be nested in there, ability id perhaps or just the link id to contain ability and paw in one) or just a singular value reflecting the item closest to the fact collection?

So should it be for a seeded fact be the source id, for a collected fact it now is the operation id or the link id, or just tag everything on there as they are available items to resolve as they all had a part in collecting the facts and searching based on different values would produce different collections that could be individually useful when analyzing things. (Aka: seeing which abilities return more facts, or what types of facts they commonly collect, or the relative data collected from one host versus another based on paw, or even two agents on the same host with different execution levels and seeing the difference in data collected by each)

Reply:
Sorry, I'm not sure I follow - I just looked through the PR, and I didn't see a single place were a paw was used as the `source` originally and now wouldn't be. Agent specific links, and the facts associated with them, use the paw as the `source`, while the `collected_by` field is used in other places (operation, etc.) to indicate the specific agent that spawned a link/fact, even if the `source` remains the same across facts. You can also figure out places that 'created' a fact by reversing each fact's list of links that generated that fact (a new feature), and checking their associated agents/paws in a pinch. 

As for your second idea, I guess we could make the source a more dynamic field that supports multiple tags, but I'm not sure the complexity that would require is worth the gain we'd get from it. The `source` field is intended to be a level of scoping for the fact, rather than a track of the true specific origin (which we have helper functions planned for). If the team wants to go that route or another one, though, I can do so.

Sorry if I misunderstood what you were going for here - if so, let me know.

Reply:
I believe you are correct, looking back through things I may have been mixing `collected_by` and `source`, as most of the older code did not set source that I have re-examined.

And if we have future functionality to resolve all the same details of data origination for all the touched components in generating a fact, then that should satisfy what I was trying to put forth in the second half of my comment on being able to draw value from the data and possibly lead to more intelligent planning and link generation.